Readme.md
# Diograph Upload

```
import { DiographUpload } from "diograph-upload"

<DiographUpload
  secrets={{ "master": "kissa123" }}
  onDioryCreated={ diory => { console.log(diory) } } />
```

## Usage / Development

```
npm install -g npx
npm install
npm start
```
Then go to: http://localhost:4204/


## Tests

```
npm test
```

## Deploy

```
npx webpack
cp app/index.html dist/index.html
# Change "../dist/bundle.js" to "bundle.js" in dist/index.html
surge ./dist diograph-upload.surge.sh
```


CHANGELOG.md
# Diograph Upload
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.0.1] 2018-10-29

Initial functionality and release


spec/diory-factory-spec.ts
import { DioryFactory } from '../app/lib/diory-factory'
import { Diory, DiographStore } from 'diograph-store'

// new Promise() requires this to work
declare var Promise: any;

describe('DioryFactory', () => {

  let genericPromise = (content={}) => {
    return new Promise(resolve => { resolve(content) })
  }

  let file = {name: "New Diory from Image file"}
  let token = "cat4321"
  // TODO: Use real Diory object instead of an empty
  let createdDiory = {}
  let uploadUrlResponseObject, exifData
  let setAuthTokenSpy, createDiorySpy, getFromEndpointSpy, extractEXIFDataSpy, uploadToS3Spy
  let createDioryFromFilePromise

  describe('createDioryFromFile', () => {

    beforeEach(() => {
      process.env.DIOGRAPH_SERVER_HOST = "http://diograph-server.test"
      // Stub all the dependencies:

      // 1. DiographStore
      setAuthTokenSpy = spyOn(DiographStore, "setAuthToken")
      createDiorySpy = spyOn(DiographStore, "createDiory").and.returnValue(genericPromise(createdDiory))

      // 2. DiographServer presigned-upload-url endpoint
      uploadUrlResponseObject = {
        "data": {
          "upload-url": "http://upload.url/",
          "public-url": "http://public.url/"
        }
      }
      getFromEndpointSpy = spyOn(DioryFactory, "getFromEndpoint").and.returnValue(genericPromise(uploadUrlResponseObject))

      // 3. EXIF data extraction
      exifData = {
        "date": "123123",
        "latitude": "67.432",
        "longitude": "32.345"
      }
      extractEXIFDataSpy = spyOn(DioryFactory, "extractEXIFData").and.returnValue(genericPromise(exifData))

      // 4. S3 upload with PUT request
      let S3ResponseObject = { "ok": true }
      uploadToS3Spy = spyOn(DioryFactory, "uploadToS3").and.returnValue(genericPromise(S3ResponseObject))


      // Generic execution
      createDioryFromFilePromise = DioryFactory.createDioryFromFile(file, token)
    })

    it('returns ', done => {
      createDioryFromFilePromise.then(diory => {
        expect(diory).toEqual(createdDiory)
        done()
      })
    })

    it('calls DiographStore.setAuthtoken with correct arguments', done => {
      createDioryFromFilePromise.then(diory => {
        expect(setAuthTokenSpy.calls.count()).toEqual(1)
        expect(setAuthTokenSpy.calls.argsFor(0)).toEqual([token])
        done()
      })
    })

    it('calls DiographStore.createDiory with correct arguments', done => {
      createDioryFromFilePromise.then(diory => {
        let expectedDioryData = exifData
        expectedDioryData["name"] = file.name
        expectedDioryData["type"] = "image"
        expectedDioryData["background"] = uploadUrlResponseObject.data["public-url"]

        expect(createDiorySpy.calls.count()).toEqual(1)
        expect(createDiorySpy.calls.argsFor(0)).toEqual([expectedDioryData])

        done()
      })
    })

    it('calls DioryFactory.getFromEndpoint with correct arguments', done => {
      createDioryFromFilePromise.then(diory => {
        expect(getFromEndpointSpy.calls.count()).toEqual(1)
        let endpoint = process.env.DIOGRAPH_SERVER_HOST + "/v1/presigned-upload-url"
        expect(getFromEndpointSpy.calls.argsFor(0)).toEqual([endpoint, token])
        done()
      })
    })

    it('calls DioryFactory.extractEXIFData with correct arguments', done => {
      createDioryFromFilePromise.then(diory => {
        expect(extractEXIFDataSpy.calls.count()).toEqual(1)
        expect(extractEXIFDataSpy.calls.argsFor(0)).toEqual([file])
        done()
      })
    })

    it('calls DioryFactory.uploadToS3 with correct arguments', done => {
      createDioryFromFilePromise.then(diory => {
        expect(uploadToS3Spy.calls.count()).toEqual(1)
        expect(uploadToS3Spy.calls.argsFor(0)).toEqual([uploadUrlResponseObject.data["upload-url"], file])
        done()
      })
    })


  })


  // extractEXIFData tests
  // TODO: Requires proper FileMockObject to run this test
  // - some kind of .html page which has an image? (see exif-js github page)
  // - load an image to a File object (one with EXIF and one without it)
  /*
  describe('DioryFactory.extractEXIFData', () => {

    it('returns null values if no EXIFData available', done => {
      let file = {name: "filename.jpg"}
      DioryFactory.extractEXIFData(file).then(exifData => {
        expect(exifData).toEqual({date: null, latitude: null, longitude: null})
        done()
      })
    })

  })
  */

})


app/lib/upload.ts
import { DiographStore, Diory } from 'diograph-store'
import { EXIF } from 'exif-js'
import * as request from "superagent"

// new Promise() requires this to work
declare var Promise: any;

export class Upload {

  static uploadFiles(event, token): any {
    let files = event.target.files
    DiographStore.setAuthToken(token);

    // TODO: Support for multiple files => some kind of iterator
    // - files is not an Array but a "FileList" which doesn't have forEach() iterator...
    let dioryData = this.generateDioryDataFromImageFile(files[0], token)
    return DiographStore.createDiory(dioryData).then(diory => {
      // return diory
      return "jeejee"
    })
  }


// ------- PRIVATE METHODS --------


// Generate diory data by getting the background and extracting the EXIF data
  static async generateDioryDataFromImageFile(file, token): Promise<object> {

    // 1. Background is the uploaded image's S3 url
    let background = await this.getBackground(file, token)

    // 2. Date, latitude & longitude are extracted from EXIF
    // TODO: Skip this if no EXIF data available => return empty object {}
    let exif = await this.extractEXIFData(file)

    // 3. Diory attributes are composed to dioryData
    return {
      name: file.name,
      type: "image",
      background: background,
      date: exif["date"],
      latitude: exif["latitude"],
      longitude: exif["longitude"]
    }
  }


// Upload to S3 and get its public url as background for the created diory

  static async getBackground(file, token) {
    // Get uploadUrl from diory-server
    let uploadUrl = await this.getUploadUrl(token)

    // Try to upload file to S3
    let S3Response
    await this.S3ManagerUpload(uploadUrl["upload-url"], file)
      .then(response => {
        S3Response = response
      })

    // Return file's public url on success
    if (S3Response.ok) {
      return uploadUrl["public-url"]
    } else {
      return "new Error(" + S3Response.body + ")"
    }
  }

  static getUploadUrl(token) {
    // TODO: Use some similar generic DiographStore / DiographServerManager method instead of duplicating it here
    return this.getFromEndpoint("http://localhost:3000/v1/presigned-upload-url", token).then(response => {
      return response.data
    })
  }

  static async S3ManagerUpload(uploadUrl, file) {
    // Upload the file to S3 via PUT request to uploadUrl
    return request
      .put(uploadUrl)
      .send(file)
      .then(response => {
        return response
      })
  }


  // TODO: Use some similar generic DiographStore / DiographServerManager method instead of duplicating it here
  private static getFromEndpoint(endpoint, token) { // query={}) {
    var promise = request
      .get(endpoint)
      // .query(query)
      .set("Accept", "application/vnd.api+json")
      .set("Authorization", token)

    return promise.then(res => {
      return res.body
    }, err => { throw err })
  }


// EXIF stuff

  static async extractEXIFData(file) {
    let self = this
    return new Promise((resolve) => {
      let exif = {}
      EXIF.getData(file, function() {
        exif["date"] = EXIF.getTag(file, "DateTimeOriginal");
        exif["latitude"] = self.toGpsDecimal(EXIF.getTag(file, "GPSLatitude"));
        exif["longitude"] = self.toGpsDecimal(EXIF.getTag(file, "GPSLongitude"));
        resolve(exif)
      });
    })
  }

  static toGpsDecimal(number) {
    return number[0].numerator + number[1].numerator /
      (60 * number[1].denominator) + number[2].numerator / (3600 * number[2].denominator);
  };

}


app/lib/diory-factory.ts
import { DiographStore, Diory } from 'diograph-store'
import { EXIF } from 'exif-js'
import * as request from "superagent"

// new Promise() requires this to work
declare var Promise: any;

export class DioryFactory {

  static async createDioryFromFile(file, token): Promise<string> {
    DiographStore.setAuthToken(token);

    // TODO: Support for multiple files => some kind of iterator
    // - files is not an Array but a "FileList" which doesn't have forEach() iterator...
    let dioryData = await this.generateDioryDataFromImageFile(file, token)
    return DiographStore.createDiory(dioryData).then(diory => {
      return diory
    })
  }


// ------- PRIVATE METHODS --------


  // Generate diory data by getting the background and extracting the EXIF data
  static async generateDioryDataFromImageFile(file, token): Promise<object> {

    // 1. Background is the uploaded image's S3 url
    let background = await this.getBackground(file, token)

    // 2. Date, latitude & longitude are extracted from EXIF
    let exif = await this.extractEXIFData(file)

    // 3. Diory attributes are composed to dioryData
    return {
      name: file.name,
      type: "image",
      background: background,
      date: exif["date"],
      latitude: exif["latitude"],
      longitude: exif["longitude"]
    }
  }


// Upload to S3 and get its public url as background for the created diory

  static async getBackground(file, token) {
    // Get uploadUrl from diory-server
    let uploadUrl = await this.getUploadUrl(token)

    // Try to upload file to S3
    let S3Response
    await this.uploadToS3(uploadUrl["upload-url"], file)
      .then(response => {
        S3Response = response
      })

    // Return file's public url on success
    return uploadUrl["public-url"]
  }

  static async uploadToS3(uploadUrl, file) {
    // Upload the file to S3 via PUT request to uploadUrl
    return request
      .put(uploadUrl)
      .send(file)
      .then(response => {
        return response
      })
  }

  static getUploadUrl(token) {
    // TODO: Use some similar generic DiographStore / DiographServerManager method instead of duplicating it here
    return this.getFromEndpoint(process.env.DIOGRAPH_SERVER_HOST + "/v1/presigned-upload-url", token).then(response => {
      return response.data
    })
  }

  // TODO: Use some similar generic DiographStore / DiographServerManager method instead of duplicating it here
  static getFromEndpoint(endpoint, token) { // query={}) {
    var promise = request
      .get(endpoint)
      // .query(query)
      .set("Accept", "application/vnd.api+json")
      .set("Authorization", token)

    return promise.then(res => {
      return res.body
    }, err => { throw err })
  }


// EXIF stuff

  static async extractEXIFData(file) {
    let self = this
    return new Promise((resolve) => {
      let exif = {}
      EXIF.getData(file, function() {
        exif["date"] = EXIF.getTag(file, "DateTimeOriginal");
        exif["latitude"] = self.toGpsDecimal(EXIF.getTag(file, "GPSLatitude"));
        exif["longitude"] = self.toGpsDecimal(EXIF.getTag(file, "GPSLongitude"));
        resolve(exif)
      });
    })
  }

  static toGpsDecimal(number) {
    if (number) {
      return number[0].numerator + number[1].numerator /
        (60 * number[1].denominator) + number[2].numerator / (3600 * number[2].denominator);
    } else {
      return null
    }
  };

}
